src/utils.rs

```diff
@@ -42,33 +34,146 @@ where
     }
 }
 
-/// Spawn a task and abort process if it panics or results in error.
-pub fn spawn_or_abort<F, T>(future: F) -> JoinHandle<T>
+pub fn spawn_monitored_with_backoff<S, F, T>(
+    future_spawner: S,
+    backoff_duration: Duration,
+) -> JoinHandle<T>
 where
     F: Future<Output = AnyhowResult<T>> + Send + 'static,
+    S: Fn() -> F + Send + Sync + 'static,
     T: Send + 'static,
 {
-    // Wrap in `AssertUnwindSafe` so we can call `FuturesExt::catch_unwind` on it.
-    let future = std::panic::AssertUnwindSafe(future);
-
     // Run task in background, returning a handle.
     tokio::spawn(async move {
-        let result = future.catch_unwind().await;
-        match result {
-            Ok(Ok(t)) => t,
-            Ok(Err(e)) => {
-                error!("Task failed: {:?}", e);
-                std::process::abort();
-            }
-            Err(e) => {
-                error!("Task panicked: {:?}", e);
-                std::process::abort();
+        loop {
+            let future = future_spawner();
+
+            // Wrap in `AssertUnwindSafe` so we can call `FuturesExt::catch_unwind` on it.
+            let future = std::panic::AssertUnwindSafe(future);
+
+            let result = future.catch_unwind().await;
+
+            match result {
+                // Task succeeded or is shutting down gracefully
+                Ok(Ok(t)) => return t,
+                Ok(Err(e)) => {
+                    error!("Task failed: {:?}", e);
```

Comment:
very minor nit, surprised that clippy doesn't catch this:
`error!("Task failed: {e:?}");`